### PR TITLE
ObjectCache with intrusive_list specialization

### PR DESCRIPTION
**Pull request by Coldwings**
ObjectCache specialization for intrusive_list value-type.

Do not need to create a intrusive_list(which is only a node pointer) as ObjectCache value in creator. Store the list as a member of ObjectCache item.

### CODE REVIEW

File: common/expirecontainer.h [hunk 1]
```diff
@@ -372,3 +373,124 @@ class ObjectCache : public ObjectCacheBase {
...
+    using Base = ObjectCacheBase;
+    using KeyedItem = Base::KeyedItem<Base::Item, KeyType>;
+
+    class Item : public KeyedItem {
+    public:
+        ListType _list;
+        using KeyedItem::KeyedItem;
+        virtual Item* construct() const override {
+            auto item = new Item(this->_key);
+            item->_obj = nullptr;
```
Comment (lihuiba) [inline common/expirecontainer.h:390]: ```item->_obj = item;``` ?

File: common/expirecontainer.h [hunk 2]
```diff
@@ -372,3 +373,124 @@ class ObjectCache : public ObjectCacheBase {
...
+public:
+    ObjectCache(uint64_t expiration) : Base(expiration, expiration / 16) {}
+    ObjectCache(uint64_t expiration, uint64_t timer_cycle)
+        : Base(expiration, timer_cycle) {}
+
+    template <typename Constructor>
+    ItemPtr ref_acquire(const InterfaceKey& key, const Constructor& ctor,
+                        uint64_t failure_cooldown = 0) {
+        auto _ctor = [&](void* item) {
+            ListType list = ctor();
```
Comment (lihuiba) [inline common/expirecontainer.h:412]: ```((Item*)item)->_list = ctor();``` ?

File: common/expirecontainer.h [hunk 3]
```diff
@@ -372,3 +373,124 @@ class ObjectCache : public ObjectCacheBase {
         return borrow(key, [] { return new ValEntity(); });
     }
 };
+
+template <typename KeyType, typename NodeType>
+class ObjectCache<KeyType, intrusive_list<NodeType>> : public ObjectCacheBase {
```
Comment (lihuiba) [inline common/expirecontainer.h:378]: inherit ObjectCache<KeyType, ValPtr>, so that Borrow can be reused.

**Review by lihuiba — APPROVED** [review]

File: common/expirecontainer.h (hunk 1 above)
Comment (Coldwings) [reply common/expirecontainer.h:390]: the `construct` member function should keep _obj as nullptr.

In `ref_acquire` implementation, item will create if not exists, and then check if _obj is empty. The real `ctor` will be called only when `_obj` equals to nullptr

File: common/expirecontainer.h (hunk 2 above)
Comment (Coldwings) [reply common/expirecontainer.h:412]: intrusive_list did not have a move constructor

File: common/expirecontainer.h (hunk 3 above)
Comment (Coldwings) [reply common/expirecontainer.h:378]: Borrow in list is not a implemention.
Those override operators are different, so as `Item` type. Inheritance may not help.

**Comment by HanLee13** [comment]
LGTM

**Comment by Coldwings** [comment]
Now list and pointer items sharing same ObjectCache code, all differences are controlled by `Item` type.

**Review by HanLee13 — APPROVED** [review]

**Review by lihuiba — APPROVED** [review]

**Pull request merged**